stl/src/tzdb.cpp

```diff
@@ -1,15 +1,345 @@
+                return nullptr;
+            }
+        } else if (U_FAILURE(_UErr) || _Result_len <= 0) {
+            _Err = __std_tzdb_error::_Icu_error;
+            return nullptr;
+        }
+
+        return _Link_buf;
+    }
+
+    template <class _Ty, class _Dtor>
+    _NODISCARD constexpr _Ty* _Report_error(_STD unique_ptr<_Ty, _Dtor>& _Info, __std_tzdb_error _Err) {
+        _Info->_Err = _Err;
+        return _Info.release();
+    }
+
+    template <class _Ty, class _Dtor>
+    _NODISCARD constexpr _Ty* _Propagate_error(_STD unique_ptr<_Ty, _Dtor>& _Info) {
+        // a bad_alloc() returns nullptr and does not set __std_tzdb_error
+        return _Info->_Err != __std_tzdb_error::_Success ? _Info.release() : nullptr;
+    }
+
+} // namespace
+
 _EXTERN_C
 
+_NODISCARD __std_tzdb_time_zones_info* __stdcall __std_tzdb_get_time_zones() noexcept {
+    // On exit---
+    //    _Info == nullptr          --> bad_alloc()
+    //    _Info->_Err == _Win_error --> failed, call GetLastError()
+    //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+    _STD unique_ptr<__std_tzdb_time_zones_info, decltype(&__std_tzdb_delete_time_zones)> _Info{
+        new (_STD nothrow) __std_tzdb_time_zones_info{}, &__std_tzdb_delete_time_zones};
+    if (_Info == nullptr) {
+        return nullptr;
+    }
+
+    if (_Acquire_icu_functions() < _Icu_api_level::_Has_icu_addresses) {
+        return _Report_error(_Info, __std_tzdb_error::_Win_error);
+    }
+
+    UErrorCode _UErr{U_ZERO_ERROR};
+    _Info->_Version = __icu_ucal_getTZDataVersion(&_UErr);
+    if (U_FAILURE(_UErr)) {
+        return _Report_error(_Info, __std_tzdb_error::_Icu_error);
+    }
+
+    _STD unique_ptr<UEnumeration, decltype(&__icu_uenum_close)> _Enum{
+        __icu_ucal_openTimeZoneIDEnumeration(USystemTimeZoneType::UCAL_ZONE_TYPE_ANY, nullptr, nullptr, &_UErr),
+        &__icu_uenum_close};
+    if (U_FAILURE(_UErr)) {
+        return _Report_error(_Info, __std_tzdb_error::_Icu_error);
+    }
+
+    // uenum_count may be expensive but is required to pre-allocate arrays.
+    int32_t _Num_time_zones = __icu_uenum_count(_Enum.get(), &_UErr);
+    if (U_FAILURE(_UErr)) {
+        return _Report_error(_Info, __std_tzdb_error::_Icu_error);
+    }
+
+    _Info->_Num_time_zones = static_cast<size_t>(_Num_time_zones);
+    _Info->_Names          = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Names == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_time_zones() cleanup is valid
+    _STD fill_n(_Info->_Names, static_cast<ptrdiff_t>(_Info->_Num_time_zones), nullptr);
+
+    _Info->_Links = new (_STD nothrow) const char*[_Info->_Num_time_zones];
+    if (_Info->_Links == nullptr) {
+        return nullptr;
+    }
+
+    // init to ensure __std_tzdb_delete_time_zones() cleanup is valid
+    _STD fill_n(_Info->_Links, static_cast<ptrdiff_t>(_Info->_Num_time_zones), nullptr);

```

Comment:
Instead of `fill_n` with `nullptr`, you can use empty braces after the array to value-initialize the elements to null, e.g. `new int[5]{}`.
The comments can be kept, and updated to say "value-init".